File: src/rnet/packets/EventPacket.js
const RNetPacket = require("./RNetPacket");
const SmartBuffer = require("smart-buffer").SmartBuffer;

class EventPacket extends RNetPacket {
    constructor() {
        super();
        this.messageType = 0x05;
        this.sourcePath = [];
    }

    getMessageBody() {
        const buffer = new SmartBuffer();

        if (this.targetPath !== undefined) {
            buffer.writeUInt8(this.targetPath.length);
            for (let i = 0; i < this.targetPath.length; i++) {
                buffer.writeUInt8(this.targetPath[i])
            }
        }
        else {
            throw new Error("this.targetPath was not set")
        }

        buffer.writeUInt8(this.sourcePath.length);
        for (var i = 0; i < this.sourcePath.length; i++) {
            buffer.writeUInt8(this.sourcePath[i])
        }

        if (this.eventID !== undefined) {
            this.writeWithInvertUInt16LE(buffer, this.eventID);
        }
        else {
            throw new Error("this.eventID has not been set.")
        }

        if (this.eventTimestamp !== undefined) {
            this.writeWithInvertUInt16LE(buffer, this.eventTimestamp);
        }
        else {
            throw new Error("this.eventTimestamp has not been set.");
        }

        if (this.eventData !== undefined) {
            this.writeWithInvertUInt16LE(buffer, this.eventData);
        }
        else {
            throw new Error("this.eventData has not been set.")
        }

        if (this.eventPriority !== undefined) {
            this.writeWithInvertUInt8(buffer, this.eventPriority);
        }
        else {
            throw new Error("this.eventPriority has not been set.");
        }

        return buffer.toBuffer();
    }

    copyToPacket(packet) {
        if (packet.messageType == this.messageType) {
            if (packet.eventID == this.eventID) {
                packet.targetControllerID = this.targetControllerID;
                packet.targetZoneID = this.targetZoneID;
                packet.targetKeypadID = this.targetKeypadID;
                packet.sourceControllerID = this.sourceControllerID;
                packet.sourceZoneID = this.sourceZoneID;
                packet.sourceKeypadID = this.sourceKeypadID;

                packet.targetPath = this.targetPath;
                packet.sourcePath = this.sourcePath;
                packet.eventID = this.eventID;
                packet.eventTimestamp = this.eventTimestamp;
                packet.eventData = this.eventData;
                packet.eventPriority = this.eventPriority;
            }
            else {
                throw new Error("Attempted to copy values to packet with different eventID");
            }
        }
        else {
            throw new Error("Attempted to copy values to packet with different messageType");
        }
    }
}

EventPacket.fromPacket = function(rNetPacket) {
    if (rNetPacket instanceof RNetPacket) {
        const eventPacket = new EventPacket();
        rNetPacket.copyToPacket(eventPacket);

        const buffer = new SmartBuffer(rNetPacket.messageBody);
        eventPacket.targetPath = [];
        {
            let length = buffer.readUInt8();
            for (let i = 0; i < length; i++) {
                eventPacket.targetPath[i] = buffer.readUInt8();
            }
        }
        eventPacket.sourcePath = [];
        {
            let length = buffer.readUInt8();
            for (let i = 0; i < length; i++) {
                eventPacket.sourcePath[i] = buffer.readUInt8();
            }
        }
        eventPacket.eventID = buffer.readUInt16LE();
        eventPacket.eventTimestamp = buffer.readUInt16LE();
        eventPacket.eventData = buffer.readUInt16LE();
        eventPacket.eventPriority = buffer.readUInt8();(0, 0, true);

        return eventPacket;
    }
    else {
        throw new TypeError("Cannot create EventPacket with anything other than RNetPacket");
    }
}

module.exports = EventPacket;

File: src/RNetPi.Core/RNet/EventPacket.cs
using System;
using System.IO;

namespace RNetPi.Core.RNet;

/// <summary>
/// Represents an RNet Event packet (MessageType = 0x05)
/// </summary>
public class EventPacket : RNetPacket
{
    public byte[] TargetPath { get; set; } = Array.Empty<byte>();
    public byte[] SourcePath { get; set; } = Array.Empty<byte>();
    public ushort EventID { get; set; }
    public ushort EventTimestamp { get; set; }
    public ushort EventData { get; set; }
    public byte EventPriority { get; set; }

    public EventPacket()
    {
        MessageType = 0x05;
    }

    protected override byte[] GetMessageBody()
    {
        using var stream = new MemoryStream();
        using var writer = new BinaryWriter(stream);

        // Write target path
        writer.Write((byte)TargetPath.Length);
        writer.Write(TargetPath);

        // Write source path
        writer.Write((byte)SourcePath.Length);
        writer.Write(SourcePath);

        // Write event fields with inversion handling
        WriteWithInvertUInt16LE(writer, EventID);
        WriteWithInvertUInt16LE(writer, EventTimestamp);
        WriteWithInvertUInt16LE(writer, EventData);
        WriteWithInvertUInt8(writer, EventPriority);

        return stream.ToArray();
    }

    public override void CopyToPacket(RNetPacket packet)
    {
        base.CopyToPacket(packet);

        if (packet is EventPacket eventPacket)
        {
            eventPacket.TargetPath = TargetPath;
            eventPacket.SourcePath = SourcePath;
            eventPacket.EventID = EventID;
            eventPacket.EventTimestamp = EventTimestamp;
            eventPacket.EventData = EventData;
            eventPacket.EventPriority = EventPriority;
        }
        else
        {
            throw new ArgumentException("Cannot copy EventPacket properties to non-EventPacket");
        }
    }

    /// <summary>
    /// Creates an EventPacket from an RNetPacket
    /// </summary>
    public static EventPacket FromPacket(RNetPacket rnetPacket)
    {
        if (rnetPacket.MessageType != 0x05)
        {
            throw new ArgumentException("Cannot create EventPacket from packet with MessageType != 0x05");
        }

        var eventPacket = new EventPacket();
        rnetPacket.CopyToPacket(eventPacket);

        if (rnetPacket.MessageBody == null || rnetPacket.MessageBody.Length == 0)
        {
            return eventPacket;
        }

        using var stream = new MemoryStream(rnetPacket.MessageBody);
        using var reader = new BinaryReader(stream);

        try
        {
            // Read target path
            var targetPathLength = reader.ReadByte();
            eventPacket.TargetPath = reader.ReadBytes(targetPathLength);

            // Read source path
            var sourcePathLength = reader.ReadByte();
            eventPacket.SourcePath = reader.ReadBytes(sourcePathLength);

            // Read event fields (little endian)
            eventPacket.EventID = reader.ReadUInt16();
            eventPacket.EventTimestamp = reader.ReadUInt16();
            eventPacket.EventData = reader.ReadUInt16();
            eventPacket.EventPriority = reader.ReadByte();
        }
        catch (EndOfStreamException)
        {
            Console.WriteLine("Warning: Incomplete EventPacket - reached end of stream while parsing");
        }

        return eventPacket;
    }
}